homeassistant/components/unifiprotect/migrate.py

```diff
@@ -81,3 +85,70 @@ async def async_migrate_buttons(
 
     if count < len(to_migrate):
         _LOGGER.warning("Failed to migate %s reboot buttons", len(to_migrate) - count)
+
+
+async def async_migrate_device_ids(
+    hass: HomeAssistant, entry: ConfigEntry, protect: ProtectApiClient
+) -> None:
+    """
+    Migrate unique IDs from {device_id}_{name} format to {mac}_{name} format.
+
+    This makes devices persist better with in HA. Anything a device is unadopted/readopted or
+    the Protect instance has to rebuild the disk array, the device IDs of Protect devices
+    can change. This causes a ton of orphaned entities and loss of historical data. MAC
+    addresses are the one persistent identifier a device has that does not change.
+
+    Added in 2022.7.0.
+    """
+
+    registry = er.async_get(hass)
+    to_migrate = []
+    for entity in er.async_entries_for_config_entry(registry, entry.entry_id):
+        parts = entity.unique_id.split("_")
+        # device ID = 24 characters, MAC = 12
+        if len(parts[0]) == 24:
+            _LOGGER.debug("Entity %s needs migration", entity.entity_id)
+            to_migrate.append(entity)
+
+    if len(to_migrate) == 0:
+        _LOGGER.debug("No entities need migration to MAC address ID")
+        return
+
+    bootstrap = await protect.get_bootstrap()
+    count = 0
+    for entity in to_migrate:
+        device = None
+        parts = entity.unique_id.split("_")
+        for model in DEVICES_THAT_ADOPT:
+            attr = f"{model.value}s"
```

Comment:
We have `async_get_devices_by_type` in `utils.py` that could be used here to avoid the `getattr`.  I've been trying to only have that happen in one place so if we change out how it works later to avoid the `getattr` we won't have multiple places to refactor.